src/tools/kotlin-schema-generator.ts

```diff
@@ -8,83 +8,102 @@
-
-  readonly fields: KotlinSchemaField[] = [];
-
-  constructor(node: SchemaNode, private forWasm: boolean) {
-    super(node);
-    this.process();
-  }
-
-  addField(opts: AddFieldOptions) {
-    if (opts.typeName === 'Reference' && this.forWasm) return;
-
-    const typeInfo = getTypeInfo({name: opts.typeName, ...opts});
-    const type = typeInfo.type;
-
-    this.fields.push({
-      ...opts,
-      ...typeInfo,
-      escaped: escapeIdentifier(opts.field),
-      nullableType: type.endsWith('?') ? type : `${type}?`
-    });
-  }
-}
-
-/**
- * Generates a schema instance for a given KotlinSchemaDescriptor
- */
-export function generateSchema(descriptor: KotlinSchemaDescriptor): string {
-  const schema = descriptor.node.schema;
+export async function generateSchema(schema: Schema): Promise<string> {
   if (schema.equals(Schema.EMPTY)) return `Schema.EMPTY`;
 
   const schemaNames = schema.names.map(n => `SchemaName("${n}")`);
   const {refinement, query} = generatePredicates(schema);
 
-  function generateFieldMap(isCollection: boolean): string {
-    const fieldPairs = descriptor.fields
-        .filter(f => !!f.isCollection === isCollection)
-        .map(({field, schemaType}) => `"${field}" to ${schemaType}`);
-    return leftPad(ktUtils.mapOf(fieldPairs, 30), 8, true);
-  }
+  const singletons: string[] = [];
+  const collections: string[] = [];
+
+  await visitSchemaFields(schema, ({isCollection, field, schemaType}) =>
+      (isCollection ? collections : singletons).push(`"${field}" to ${schemaType}`));
 
   return `\
 Schema(
     setOf(${ktUtils.joinWithIndents(schemaNames, {startIndent: 8})}),
     SchemaFields(
-        singletons = ${generateFieldMap(false)},
-        collections = ${generateFieldMap(true)}
+        singletons = ${leftPad(ktUtils.mapOf(singletons, 30), 8, true)},
+        collections = ${leftPad(ktUtils.mapOf(collections, 30), 8, true)}
     ),
-    ${quote(descriptor.node.hash)},
+    ${quote(await schema.hash())},
     refinement = ${refinement},
     query = ${query}
 )`;
 }
 
+interface SchemaField {
+  field: string;
+  isCollection: boolean;
+  schemaType: string;
+}
+
+async function visitSchemaFields(schema: Schema, visitor: (field: SchemaField) => void) {
+  for (const [field, descriptor] of Object.entries(schema.fields)) {
+    switch (descriptor.kind) {
+      case 'schema-collection':
+        visitor({field, isCollection: true, schemaType: await getSchemaType(field, descriptor.schema)});
+        break;
+      case 'schema-primitive':
+      case 'kotlin-primitive':
+      case 'schema-reference':
+      case 'schema-nested':
+        visitor({field, isCollection: false, schemaType: await getSchemaType(field, descriptor)});
+        break;
+      case 'schema-ordered-list':
+        visitor({field, isCollection: false, schemaType: await getSchemaType(field, {
+          ...descriptor,
+          innerType: await getSchemaType(field, descriptor.schema)
+        })});
+        break;
+      default:
```

Comment:
How are `schema-union`, `schema-tuple`, `schema-inline` handled? It seems like in `EntityDescriptorBase`, these also were also underspecified -- so, likely it would be ok to skip for now.

Reply:
If I remember correctly, tuples are implemented at a higher level than a schema instance -- right?

Reply:
there are tuples at both levels.

Currently we don't handle all these variants in schema generation. This PR does not change that.